features/achordion.c

```diff
@@ -137,6 +147,15 @@ bool process_achordion(uint16_t keycode, keyrecord_t* record) {
       tap_hold_record.event.pressed = false;
       // Plumb hold release event.
       recursively_process_record(&tap_hold_record, STATE_RELEASED);
+    } else if (!pressed_another_key_before_release) {
```

Comment:
Nice, this is a yet more elegant solution to supporting retro tapping.

A small issue is this logic briefly presses and releases the hold function even for non-retro tap keys, where it should rather be a no-op. QMK is known to have an edge-case bug or two where quick events are involved (discussed in e.g. https://github.com/getreuer/qmk-keymap/issues/48 and https://github.com/qmk/qmk_firmware/pull/19405). It's plausible this quick hold press and release increases the occurrence of triggering such bugs.

To be on the safe side, how about enclose this else if block in `#if defined(RETRO_TAPPING) || defined(RETRO_TAPPING_PER_KEY)`. WDYT?

Reply:
I didn't know about those issues. Allowing this behaviour to be opt-in makes sense.

As to how to gate it, one thing to note is that this is not specific to retro tapping, auto shift, retro shift, retro tap etc all work with this as well. In addition to this, one of my motivations for the newer version was that I had a "fake" layer tap key where I intercept the hold action and send a custom macro instead of changing the layer, as recommended in QMK docs. Achordion was interfering with the hold and it was causing the hold action to fire later than necessary. The only way I can think of to support such keycodes is allowing the user to implement a function to tell Achordion for which keys the user wants this behaviour, and then Achordion checking the keycode with that function. I can implement that, however I'm not sure if it will be too heavy-duty for something that's not directly supported. What do you think?

Reply:
These are great points! You have convinced me =) It is valuable to interoperate with auto shift, retro shift, plus custom behaviors like that tap/long-press trick. Let's move forward with the PR as is.